### PR TITLE
add mlnet.cli to automl benchmark

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
[mlnet](https://github.com/dotnet/machinelearning) is a C# machinelearning library, [mlnet.cli](https://docs.microsoft.com/en-us/dotnet/machine-learning/automate-training-with-cli) is an automl cli tool implemented by mlnet.

#265

### CODE REVIEW

**Comment by PGijsbers** [comment]
Thank you for contributing to the benchmark project! Great to see a new tool (and language) represented 🎉 
Please indicate when this PR is ready for review by leaving a request through Github (through the "request review" on the right). If there are any questions, don't hesitate to ask.

Don't forget to update the documentation as well! We want to keep a little description with links to relevant literature/docs of each framework in at [docs/automl_overview.md](https://github.com/openml/automlbenchmark/blob/master/docs/automl_overview.md).

**Comment by LittleLittleCloud** [comment]
Hi @PGijsbers 
Thanks, I'm really excited to be able to contribute a new framework to automl benchmark.

Currently all the tests for local mode have been passed, could you show me how to test in docker and aws mode? Specifically for aws mode,, do I need to open a aws account for that

**Comment by PGijsbers** [comment]
If you don't have access/credits for `aws` you feel comfortable using, then we can do a test run. 
As long as `python runbenchmark.py my_framework validation -m docker -s force` runs successfully, it should be good to go (if you're on Windows you might need to wait for #269 to merge).

File: frameworks/MLNet/exec.py
Comment (sebhrusen) [inline frameworks/MLNet/exec.py:43]: does MLNET support ARFF files?
If so you can access the files path directly using `dataset.train.path`, as well as training information like `dataset.target`, `dataset.predictors`... 

Those work not only with OpenML but also when using a custom file (which could be a csv...).
So it's much preferable to pass the files directly to MLNet and avoid all those data conversion that can be very expensive for larger datasets.
If MLNet doesn't support `arff` files, but only `csv`, I would suggest to do the conversion only for those.

File: frameworks/MLNet/exec.py
Comment (sebhrusen) [inline frameworks/MLNet/exec.py:29]:
```suggestion
    # from frameworks.shared.callee import output_subdir
    temp_output_folder = output_subdir(name, config)
```
this standardizes folder structure and ensures that you get a log file for each task/fold.

File: frameworks/MLNet/exec.py
Comment (sebhrusen) [inline frameworks/MLNet/exec.py:45]: feel free to use f-strings everywhere if you think it improves readability, also consecutive strings are automatically concatenated in Python:
```python
cmd = f"{mlnet} {sub_command}"\
      f" -- dataset {train_dataset} --test-dataset {test_dataset}"
```
or you can use a single multiline:
```python
cmd = f"""{mlnet} {sub_command}
--dataset {train_dataset} --test-dataset {test_dataset}
--train-time {train_time_in_seconds}
"""
```

File: frameworks/MLNet/exec.py
Comment (sebhrusen) [inline frameworks/MLNet/exec.py:106]: can't you use `save_predictions` like you did for classification?
This guarantees better compatibility for various kind of datasets.

**Review by sebhrusen — CHANGES_REQUESTED** [review]

File: resources/frameworks.yaml
Comment (LittleLittleCloud) [reply resources/frameworks.yaml:160]: ### Resolved

File: .gitignore
Comment (LittleLittleCloud) [reply .gitignore:47]: ### Resolved

File: frameworks/MLNet/setup.sh
Comment (LittleLittleCloud) [reply frameworks/MLNet/setup.sh:5]: ### Resolved

File: frameworks/MLNet/setup.sh
Comment (LittleLittleCloud) [reply frameworks/MLNet/setup.sh:13]: ### Resolved

File: frameworks/MLNet/exec.py
Comment (LittleLittleCloud) [reply frameworks/MLNet/exec.py:116]: ### Resolved

File: frameworks/MLNet/exec.py
Comment (LittleLittleCloud) [reply frameworks/MLNet/exec.py:15]: ### Resolved

File: frameworks/MLNet/exec.py
Comment (LittleLittleCloud) [reply frameworks/MLNet/exec.py:43]: Unfortunately, MLNet.cli only support csv currently, so I guess the only thing I can do here is to avoid conversion when dataset's format is csv? Could you point out the code/property I can use to determine the file format?

File: frameworks/MLNet/exec.py
Comment (LittleLittleCloud) [reply frameworks/MLNet/exec.py:29]: Resolved

File: frameworks/MLNet/exec.py
Comment (LittleLittleCloud) [reply frameworks/MLNet/exec.py:45]: ### Resolved

File: frameworks/MLNet/exec.py
Comment (LittleLittleCloud) [reply frameworks/MLNet/exec.py:106]: ### Resolved

**Comment by LittleLittleCloud** [comment]
hi @sebhrusen, @PGijsbers Any follow up to this PR :-)

**Comment by sebhrusen** [comment]
@LittleLittleCloud sorry, very very busy lately, I'll have a look at it from Wed.

File: frameworks/MLNet/exec.py
Comment (sebhrusen) [reply frameworks/MLNet/exec.py]: you could check the extension of `dataset.train.path` for example

File: frameworks/MLNet/exec.py
Comment (sebhrusen) [inline frameworks/MLNet/exec.py:29]: there are lot of things going into this temp folder, and as it is inside the output dir, all the files created there get uploaded to s3 at the end of the benchmark when it runs on aws, and at the end, downloaded to the machine that started the benchmark, so we want to avoid downloading unnecessary artifacts.

I would recommend to create this tmp folder using:
```
with tempfile.TemporaryDirectory() as tmpdir:
     ...
 ```
 to ensure that the content gets deleted at the end.
 
 if you want to keep the logs, or various artifacts, we accept a param named `_save_artifacts` (see `autosklearn` for an example), for example a user creating its custom framework:
 ```
 myMLNet:
    extends: MLNet
    params:
       _save_artifact = ['logs', 'models']
```
would be able to retrieve the framework logs and the serialized models.
In your case it would mean sth like:
```python
artifacts = config.framework_params.get('_save_artifacts', [])
log_dir = output_subdir('logs', config=config) if 'logs' in artifacts else os.path.join(tmpdir, 'logs')
mlnet_output_dir = output_subdir('models', config=config) if 'models' in artifacts else os.path.join(tmpdir, 'models')
```
this way the `config.output_dir` is not polluted if the extra artifacts were not requested by the user.

File: frameworks/MLNet/exec.py
Comment (sebhrusen) [inline frameworks/MLNet/exec.py:46]: a simpler alternative probably to this:
- if `dataset.train.path` extension  is `.csv` then it is already a `csv` file and the `label` column is `dataset.target.name`.
- if `dataset.train.path` extension  is `.arff` then you could use the fact that ARFF format is a superset of CSV, which would allow you to convert it to csv format without having to load it in memory: I can help providing a utility function for this if you want. We know the column names and we know that after line `@DATA` it's basically a csv file.

**Review by sebhrusen — COMMENTED** [review]
@LittleLittleCloud, thanks for the previous changes.
Now adding new comments focusing on standardizing the integration logic.

Otherwise, I could test the benchmark locally, the setup on macOS was smooth and the test run worked great!

Did you try to run the `validation` to see if it handles some weird datasets correctly?
For a complete validation, I recommend:
```python runbenchmark.py mlnet validation test -m docker -s force```

File: resources/frameworks.yaml
Comment (LittleLittleCloud) [reply resources/frameworks.yaml:159]: `MLNet` is still in pre-release state so we don't have a `stable` version yet. Can I treat `stable` the same with `latest` for now.

File: frameworks/MLNet/exec.py
Comment (LittleLittleCloud) [reply frameworks/MLNet/exec.py:46]: I can make the change to avoid creating a new csv file from original dataset if dataset's extension is `.csv`. However, if the original dataset's extension is `.arff`, since MLNet can't load .arff dataset, we still need to save the converted dataset into a separate copy so it will still be loaded into memory. 

However, if you just want to avoid creating `Dataframe` because it's a null operation for MLNet and it's very costy for large dataset, and you want to replace it with a cheaper operation. I can follow what you suggested and it would be great if you can provide function on how to do the conversion

File: frameworks/MLNet/exec.py
Comment (LittleLittleCloud) [reply frameworks/MLNet/exec.py:29]: Sure

File: frameworks/MLNet/exec.py
Comment (LittleLittleCloud) [reply frameworks/MLNet/exec.py:11]: I'm not quite following, do you want me to reorder the import statement?

File: frameworks/MLNet/exec.py
Comment (LittleLittleCloud) [reply frameworks/MLNet/exec.py:29]: ### Resolved

File: resources/frameworks.yaml
Comment (LittleLittleCloud) [reply resources/frameworks.yaml]: ### Resolved

File: frameworks/MLNet/setup.sh
Comment (LittleLittleCloud) [reply frameworks/MLNet/setup.sh:27]: ### Resolved

File: frameworks/MLNet/exec.py
Comment (sebhrusen) [reply frameworks/MLNet/exec.py]: yes, grouping + reordering please, thank you.

File: frameworks/MLNet/exec.py
Comment (sebhrusen) [reply frameworks/MLNet/exec.py]: > However, if you just want to avoid creating Dataframe because it's a null operation for MLNet and it's very costy for large dataset, and you want to replace it with a cheaper operation. I can follow what you suggested and it would be great if you can provide function on how to do the conversion

Yes, that's what I mean, I'll provide the function soon.

File: frameworks/MLNet/exec.py
Comment (LittleLittleCloud) [inline frameworks/MLNet/exec.py:53]: when it's 'csv', simply use the dataset path to load data.
when it's 'arff', use `dataset.train.data` to avoid loading data twice.

File: frameworks/MLNet/exec.py
Comment (LittleLittleCloud) [inline frameworks/MLNet/exec.py:64]: After comparing between 1) directly saving ndarray into csv and 2) create df and call `df.to_csv` instead, it appears that the 2) is still a better option because it's much faster and by avoiding creating a variable to save df, the memory will be released immediately after `df.to_csv` get called. Because `dataset.train.data` is a lazy property, dataset will only be loaded into memory when get called and will only be loaded once so perf issue here should be small.

**Review by sebhrusen — APPROVED** [review]
Looks good to me: tested the docker image with validation benchmark, everything worked perfectly!
Thanks for this contribution.
Please address the small comment in `frameworks.yaml` and then I'll merge.

File: resources/frameworks.yaml
Comment (LittleLittleCloud) [reply resources/frameworks.yaml]: ## Resolved

**Pull request merged**
